### PR TITLE
pfn_notify callback added for clBuildPorgram, related test added and -Werror test revised

### PR DESCRIPTION
**Pull request by ardacoskunses**
pfn_notify callback added for clBuildProgram, and related test added to clBuildProgram tests.

OCL1.2 spec says:
"If pfn_notify is not NULL, clBuildProgram does not need to wait for the build to complete and can return immediately once the build operation can begin."
This feature still missing.

Also -Werror test revised to test full functionality: warning in the kernel build turned to error and checked in the test.

### CODE REVIEW

File: lib/CL/clBuildProgram.c [hunk 1]
```diff
@@ -204,6 +205,21 @@ CL_API_SUFFIX__VERSION_1_0
 
   POCL_LOCK_OBJ(program);
 
+  if(pfn_notify)
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Please add white space after if. Check https://www.gnu.org/prep/standards/html_node/Formatting.html

File: lib/CL/clBuildProgram.c [hunk 2]
```diff
@@ -495,6 +511,11 @@ CL_API_SUFFIX__VERSION_1_0
   program->build_status = CL_BUILD_SUCCESS;
   POCL_UNLOCK_OBJ(program);
 
+  /* callback freed in ProgramRelease */
+  if(program->buildprogram_callbacks)
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Ditto.

File: lib/CL/clBuildProgram.c [hunk 3]
```diff
@@ -520,6 +541,12 @@ CL_API_SUFFIX__VERSION_1_0
    * cause a double free. */
 
 ERROR:
+  if(program->buildprogram_callbacks)
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Here also.

File: lib/CL/clBuildProgram.c [hunk 4]
```diff
@@ -204,6 +205,21 @@ CL_API_SUFFIX__VERSION_1_0
 
   POCL_LOCK_OBJ(program);
 
+  if(pfn_notify)
+    {
+      callback = (build_program_callback_t*) malloc (sizeof (build_program_callback_t));
+      if (callback == NULL)
+      {
+        POCL_UNLOCK_OBJ(program);
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Also white space is needed before the argument list (in the GNU C style).

File: lib/CL/clBuildProgram.c [hunk 5]
```diff
@@ -520,6 +541,12 @@ CL_API_SUFFIX__VERSION_1_0
    * cause a double free. */
 
 ERROR:
+  if(program->buildprogram_callbacks)
+    {
+      program->buildprogram_callbacks->callback_function (program,
+                         program->buildprogram_callbacks->user_data);
+      POCL_MEM_FREE(program->buildprogram_callbacks);
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Ditto.

File: lib/CL/clReleaseProgram.c [hunk 6]
```diff
@@ -59,6 +59,9 @@ POname(clReleaseProgram)(cl_program program) CL_API_SUFFIX__VERSION_1_0
           k->program = NULL;
         }
 
+      if(program->buildprogram_callbacks)
```
Comment (pjaaskel) [inline lib/CL/clReleaseProgram.c]: Ditto (twice).

File: lib/CL/clReleaseProgram.c [hunk 7]
```diff
@@ -59,6 +59,9 @@ POname(clReleaseProgram)(cl_program program) CL_API_SUFFIX__VERSION_1_0
...
+        POCL_MEM_FREE(program->buildprogram_callbacks);
```
Comment (pjaaskel) [inline lib/CL/clReleaseProgram.c]: Isn't this a linked list?

File: lib/CL/clBuildProgram.c [hunk 8]
```diff
@@ -495,6 +511,11 @@ CL_API_SUFFIX__VERSION_1_0
   program->build_status = CL_BUILD_SUCCESS;
   POCL_UNLOCK_OBJ(program);
 
+  /* callback freed in ProgramRelease */
+  if(program->buildprogram_callbacks)
+      program->buildprogram_callbacks->callback_function (program,
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Shouldn't it traverse the linked list?

**Comment by ardacoskunses** [comment]
Agreed all the comments and updated the PR accordingly.

Linked list implementation is not necessary for clBuildProgram pfn_notify so I changed to single callback.
**Note on this:** all implemented callback functions in pocl are linkedlist (clSetMemObjectDestructorCallback, and clEventCallback) OCL1.1 spec explains their usage scenarios as multiple callbacks, so it makes sense using linked list for them. 
My understanding from spec clBuildProgram callback does **not** have to be linked list. 
Please let me know if otherwise.

File: include/pocl.h [hunk 9]
```diff
@@ -58,6 +58,14 @@ struct _mem_destructor_callback
   mem_destructor_callback_t *next;
 };
 
+typedef struct _build_program_callback build_program_callback_t;
+/* represents a build program callback */
```
Comment (pjaaskel) [inline include/pocl.h]: This comment is useless as it basically repeats the struct name.

**Pull request merged**